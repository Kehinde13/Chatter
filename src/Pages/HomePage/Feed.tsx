import React from 'react'
import { useOutletContext } from 'react-router-dom'

function Feed() {
  const [showSideBar] = useOutletContext()
  return (
    <div  className={`p-5 sm:block ${showSideBar ? "hidden" : "ml-3"}`}>
      <div>
        ioerbce sdniocerve jhiosd sdochuc jkasducysc jac7ceyuw
      </div>
    </div>
  )
}

export default Feed