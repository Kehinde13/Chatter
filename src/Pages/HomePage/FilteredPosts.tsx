import PostsCard from './Feed/PostsCard';
import Loading from '../../components/Loading';
import { useParams } from 'react-router-dom';
import { Blog } from '../../Context/Context';
import { Post } from '../../hooks/GetPosts';



function FilteredPosts() {
    const { tag } = useParams<{ tag?: string  }>();
    const { posts, postLoading } = Blog();

    if (!tag) {
      return (
        <section className="sm:mx-3 w-[80%]">
          <div>
            <h3 className="text-3xl pb-6 border-b border-purple-500 mb-[3rem] ">
              There are no tag specified
            </h3>
          </div>
        </section>
      );
    }
  
    const filteredData = posts.filter((post: Post) => post.tags.includes(tag));
  
    return (
      <section className="sm:mx-3 w-[80%]">
        <div>
          <h3 className="text-3xl pb-6 border-b border-purple-500 mb-[3rem] ">
            {filteredData.length
              ? "Your Filtered Posts "
              : "There are no post with this tag"}
          </h3>
          {postLoading ? (
            <Loading />
          ) : (
            <div className="lg:max-w-[60%] mx-10 flex flex-col gap-[2rem]">
              {filteredData &&
                filteredData.map((post: Post, i: number) => <PostsCard post={post} key={i} />)}
            </div>
          )}
        </div>
      </section>
    );
}

export default FilteredPosts