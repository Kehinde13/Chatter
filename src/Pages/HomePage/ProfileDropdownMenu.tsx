import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuPortal,
  DropdownMenuSeparator,
  DropdownMenuSub,
  DropdownMenuSubContent,
  DropdownMenuSubTrigger,
  DropdownMenuTrigger,
} from "../../components/shadcn/dropdown-menu";
import profileImg from "../../assets/profile.jpg";
import { Blog } from "../../Context/Context";
import { Link, useNavigate } from "react-router-dom";
import { auth } from "../../Auth/firebase";
import { signOut } from "firebase/auth";
import { toast } from "react-toastify";
import { FirebaseError } from "firebase/app";

type DropdownMenuProps = {
  userImg?: string;
};

export function UserDropdownMenu({ userImg }: DropdownMenuProps) {
  const { currentUser } = Blog();
  const navigate = useNavigate();

  const logout = async () => {
    try {
      await signOut(auth);
      navigate("/");
      toast.success("User has been logged out");
    } catch (error: unknown) {
      if (error instanceof FirebaseError) {
        toast.error(error.message);
      } else {
        toast.error("An error occurred");
      }
    }
  };

  const user = currentUser as { uid?: string; email?: string };

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <img
          src={userImg || profileImg}
          alt="user profile img"
          className="w-[7rem] h-[3rem] md:w-[3rem] md:h-[3rem]  object-cover rounded-full self-center"
        />
      </DropdownMenuTrigger>
      <DropdownMenuContent className="w-56">
        <DropdownMenuLabel>My Account</DropdownMenuLabel>
        <DropdownMenuSeparator />
        <DropdownMenuGroup>
          <DropdownMenuItem>
            <Link to={`profile/${user.uid}`}>Profile</Link>
          </DropdownMenuItem>
          <DropdownMenuItem>
            Billing
          </DropdownMenuItem>
          <DropdownMenuItem>
            Settings
          </DropdownMenuItem>
        </DropdownMenuGroup>
        <DropdownMenuSeparator />
        <DropdownMenuGroup>
          <DropdownMenuItem>Team</DropdownMenuItem>
          <DropdownMenuSub>
            <DropdownMenuSubTrigger>Invite users</DropdownMenuSubTrigger>
            <DropdownMenuPortal>
              <DropdownMenuSubContent>
                <DropdownMenuItem>Email</DropdownMenuItem>
                <DropdownMenuItem>Message</DropdownMenuItem>
                <DropdownMenuSeparator />
                <DropdownMenuItem>More...</DropdownMenuItem>
              </DropdownMenuSubContent>
            </DropdownMenuPortal>
          </DropdownMenuSub>
          <DropdownMenuItem>
            New Team
          </DropdownMenuItem>
        </DropdownMenuGroup>
        <DropdownMenuSeparator />
        <DropdownMenuItem>GitHub</DropdownMenuItem>
        <DropdownMenuItem>Support</DropdownMenuItem>
        <DropdownMenuItem disabled>API</DropdownMenuItem>
        <DropdownMenuSeparator />
        <DropdownMenuItem onClick={logout} className="text-red-500">Log out</DropdownMenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}
