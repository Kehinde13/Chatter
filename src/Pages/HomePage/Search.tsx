import { FaMagnifyingGlass } from "react-icons/fa6";
import { MdOutlineCancel } from "react-icons/md";
import { useState } from "react";
import { Blog } from "../../Context/Context";
import { useNavigate } from "react-router-dom";

interface Post {
  id: string;
  title: string;
  desc: string;
}

interface Props {
  searchBar: boolean;
  toggleSearchBar: (searchBar: boolean) => void;
}

function Search({ searchBar, toggleSearchBar }: Props) {
  const [search, setSearch] = useState<string>("");
  const { posts } = Blog();
  const navigate = useNavigate();

  const searchData: Post[] | undefined = posts && posts.filter((post: any) =>
    post.title.toLowerCase().includes(search.toLowerCase())
  );

  return (
    <div className="relative">
      <div
        className={`sm:flex sm:mb-0 items-center gap-2 bg-gray-100 px-2 rounded-full relative z-10 sm:w-[300px]
     duration-300  ${searchBar ? "flex " : "mb-[200px] w-[20%]"}`}
      >
        <button className="text-xl text-gray-400">
          <FaMagnifyingGlass />
        </button>
        <input
          className="bg-transparent outline-none py-[0.7rem] text-sm w-full"
          type="text"
          placeholder="Search Chatter"
          value={search}
          onChange={(e) => setSearch(e.target.value)}
        />
        {searchBar && (
          <MdOutlineCancel
            className="text-slate-800 text-2xl"
            onClick={() => toggleSearchBar(false)}
          />
        )}
      </div>
      {search !== "" && (
        <div className="absolute right-0 left-0 top-full bg-white shadow rounded-md">
          {searchData && searchData.length > 0 ? (
            <>
              {searchData.map((post: Post, i: number) => (
                <div
                  key={i}
                  onClick={() => {
                    navigate(`/HomePage/SinglePost/${post.id}`);
                    setSearch("");
                  }}
                  className="p-2 border-b border-gray-300 hover:bg-gray-100 cursor-pointer"
                >
                  <h2 className="line-clamp-1 capitalize text-sm font-bold">
                    {post.title}
                  </h2>
                  <div
                    className="text-xs text-gray-500 line-clamp-2"
                    dangerouslySetInnerHTML={{ __html: post.desc }}
                  />
                </div>
              ))}
            </>
          ) : (
            <p className="text-sm text-gray-500 p-3">No Post Found</p>
          )}
        </div>
      )}
    </div>
  );
}

export default Search;
