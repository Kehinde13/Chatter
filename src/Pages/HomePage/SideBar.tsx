import { Link } from "react-router-dom";
import Follow from "./Features/Follow";

interface SidebarProps {
  showSideBar: boolean;
  setShowSideBar: (showSideBar: boolean) => void; 
}

function SideBar({ showSideBar, setShowSideBar }: SidebarProps) {
  const topics = [
    "Programming",
    "Data Science",
    "Machine Learning",
    "Technology",
    "Politics",
    "markdown",
    "Updates",
  ];

  const handleClick = () => {
    setShowSideBar(false); 
  };

  return (
    <aside
      className={`sticky h-screen sm:overflow-y-scroll top-0  pb-4 flex flex-col sm:w-[20%] sm:ml-0 pr-10 sm:border-r-2
                   duration-500 ${
                     showSideBar ? "ml-1 w-full" : "ml-[-280px] border-r-0"
                   }`}
    >
      <h1 className="font-bold text-lg ml-2">Following</h1>

      <div className="ml-1">
        <Follow showSideBar={showSideBar} setShowSideBar={setShowSideBar}/>
      </div>

      <div className="flex gap-2">
        <h1 className="font-bold text-lg ml-2 text-center">Discover</h1>
      </div>
      <ul className="ml-5">
        {topics.map((topic, i) => (
          <li key={i} className="p-2">
            <Link to={`/HomePage/FilteredPosts/${topic}`} onClick={handleClick}>
              {topic}
            </Link>
          </li>
        ))}
      </ul>
    </aside>
  );
}

export default SideBar;
