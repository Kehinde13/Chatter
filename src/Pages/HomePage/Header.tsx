import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import profileImg from "../../assets/profile.jpg";
import { useState } from "react";
import logo from "../../assets/CHATTER.png";
import { Blog } from "../../Context/Context";
import { Link, useParams } from "react-router-dom";
import Loading from "../../components/Loading";
import Modal from "../../components/Modal";
import UserModal from "../UserModal";
import Search from "./Search";

function Header() {
  const { currentUser, users, userLoading } = Blog();
  const [searchBar, setSearchBar] = useState(false);
  const [modal, setModal] = useState(false);
  const { userId } = useParams();

  const toggleSearchBar = () => {
    setSearchBar(!searchBar);
  };

  const toggleModal = () => {
    setModal(!modal);
  };

  const currentUserData = users?.find((user) => user.id === currentUser?.uid);

  return (
    <header>
      {userLoading && <Loading />}
      <div className="flex gap-2 justify-between mx-3 md:mx-5 items-center h-[70px] h-shadow-lg">
        <Link to="/HomePage">
          <img
            src={logo}
            className={`${searchBar ? "w-[80px]" : "w-[120px]"}`}
          />
        </Link>
        
        <Search searchBar={searchBar} toggleSearchBar={toggleSearchBar} />
        
        {!searchBar && (
          <div className="flex gap-5">
            <FontAwesomeIcon
              icon="fa-solid fa-magnifying-glass"
              className="self-center sm:hidden block"
              onClick={toggleSearchBar}
            />
            <div className="flex items-center relative" onClick={toggleModal}>
              <img
                className="sm:w-[2.5rem] sm:h-[2.5rem] w-[70px] h-[35px] object-cover rounded-full cursor-pointer"
                src={currentUserData?.userImg || profileImg}
                alt="profile-img"
              />

              <Modal modal={modal} toggleModal={toggleModal}>
                <div
                  className={`${
                    modal ? "visible opacity-100%" : "invisible opacity-0"
                  } transition-all duration-100`}
                >
                  <UserModal toggleModal={toggleModal} />
                </div>
              </Modal>
            </div>
          </div>
        )}
      </div>
    </header>
  );
}

export default Header;
