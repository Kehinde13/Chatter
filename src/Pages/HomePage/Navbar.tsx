import { FaMagnifyingGlass } from "react-icons/fa6";
import { useState } from "react";
import { Blog } from "../../Context/Context";
import { Link } from "react-router-dom";
import logo from "../../assets/CHATTER.png";
import Search from "./Search";
import Loading from "../../components/Loading";
import { UserDropdownMenu } from "./ProfileDropdownMenu";
import { ModeToggle } from "../../components/shadcn/mode-toggle";

interface UserData {
  id: string;
  userImg?: string;
}

function Header() {
  const { currentUser, users, userLoading } = Blog();
  const [searchBar, setSearchBar] = useState<boolean>(false);

  const toggleSearchBar = (): void => {
    setSearchBar((prevState) => !prevState);
  };

  const currentUserData: UserData | undefined = users?.find(
    (user) => user.id === currentUser?.uid
  );

  return (
    <div>
      {userLoading && <Loading />}
      <div className="flex gap-2 justify-between px-2 md:px-5 items-center h-[70px] shadow-lg mb-5">
        <Link to="/homepage">
          <img
            src={logo}
            className={`${searchBar ? "w-[100px]" : "w-[200px] md:w-[100px]"}`}
            alt="logo"
          />
        </Link>

        
        <Search searchBar={searchBar} toggleSearchBar={toggleSearchBar} />
        

        {!searchBar && (
          <div className="flex gap-5 justify-end">
            <FaMagnifyingGlass
              className="self-center sm:hidden block text-5xl"
              onClick={toggleSearchBar}
            />

            <UserDropdownMenu userImg={currentUserData?.userImg} />
            <ModeToggle />
          </div>
        )}
      </div>
    </div>
  );
}

export default Header;
