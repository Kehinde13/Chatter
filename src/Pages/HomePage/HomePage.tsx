import { Outlet } from "react-router-dom";
import Header from "./Header";
import SideBar from "./SideBar";
import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

function HomePage() {
  const [showSideBar, setShowSideBar] = useState<boolean>(false);

  const toggleSideBar = () => {
    setShowSideBar(!showSideBar);
    console.log(showSideBar);
  };

  return (
    <div className='dark:bg-slate-800 dark:text-white '>
      <Header />
      <div className="flex">
        <button
          className={`border w-8 py-1 absolute sm:hidden rounded-full top-64 z-50 ${
            showSideBar ? "right-0" : "left-0"
          }`}
          onClick={toggleSideBar}
        >
          <FontAwesomeIcon
            icon={
              showSideBar ? "fa-solid fa-angle-left" : "fa-solid fa-angle-right"
            }
          />
        </button>
        <SideBar showSideBar={showSideBar} setShowSideBar={setShowSideBar} />
        <Outlet context={[showSideBar]} />
      </div>
    </div>
  );
}

export default HomePage;
