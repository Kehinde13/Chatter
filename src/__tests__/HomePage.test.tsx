import { render } from '@testing-library/react';
import HomePage from '../Pages/HomePage/HomePage'; 
import { MemoryRouter } from 'react-router-dom';

jest.mock('../Pages/HomePage/Navbar', () => () => <div>Mocked Header</div>);
jest.mock('../Pages/HomePage/SideBar', () => () => <div>Mocked SideBar</div>);
/* jest.mock('react-router-dom', () => ({
  Outlet: () => <div>Mocked Outlet</div>,
})); */

test('HomePage matches snapshot', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <HomePage />
    </MemoryRouter>
  );
  
  expect(asFragment()).toMatchSnapshot();
});
