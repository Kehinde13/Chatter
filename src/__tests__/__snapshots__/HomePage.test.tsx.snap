// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="homepage"
  >
    <div>
      Mocked Header
    </div>
    <div
      class="md:grid grid-cols-8"
    >
      <div>
        Mocked SideBar
      </div>
    </div>
  </div>
</DocumentFragment>
`;
